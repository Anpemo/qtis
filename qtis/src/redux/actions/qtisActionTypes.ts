export default {
  USER_REGISTER: 'USER_REGISTER',
  PRODUCTS_LIST: 'PRODUCTS_LIST',
  SINGLE_PRODUCT: 'SINGLE_PRODUCT',
  CREATE_PRODUCT: 'CREATE_PRODUCT',
  REVIEWS_LIST: 'REVIEWS_LIST'
}
