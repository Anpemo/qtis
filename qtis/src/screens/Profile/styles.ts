import { StyleSheet } from 'react-native'
import { COLORS, SIZES, SHADOW, BORDER } from '../../../constants'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    backgroundColor: COLORS.white,
    width: '100%'
  },
  scroll: {
    flex: 1,
    backgroundColor: COLORS.white,
    width: '100%',
    alignContent: 'center'
  },
  header: {
    flex: 0.5,
    resizeMode: 'cover',
    alignItems: 'flex-end',
    justifyContent: 'center',
    width: '100%'
  },
  pictureBox: {
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'center',
    height: 165,
    width: 165,
    borderRadius: 100,
    backgroundColor: COLORS.white,
    marginBottom: 10
  },
  userPicture: {
    height: 125,
    width: 125,
    resizeMode: 'cover',
    borderRadius: 100,
    backgroundColor: COLORS.white
  },
  userPictureText: {
    fontFamily: 'MontserratBold'
  },
  cameraButtonContainer: {
    backgroundColor: COLORS.white,
    position: 'absolute',
    bottom: -5,
    right: -10,
    padding: 7,
    borderRadius: 50
  },
  body: {
    flex: 1,
    width: '100%',
    alignItems: 'center',
    marginBottom: 40
  },
  userInformation: {
    alignSelf: 'center',
    justifyContent: 'center',
    width: '95%',
    marginTop: 10,
    borderRadius: SIZES.squareRadius,
    ...SHADOW
  },
  userName: {
    fontSize: SIZES.p20,
    color: COLORS.black,
    alignSelf: 'center',
    fontFamily: 'MontserratBold'
  },
  userData: {
    fontSize: SIZES.p20,
    color: COLORS.black,
    alignSelf: 'center',
    fontFamily: 'Montserrat',
    marginTop: 5
  },
  updateButton: {
    margin: 5,
    width: SIZES.buttonwidth * 0.8,
    height: SIZES.buttonheight * 0.9,
    borderRadius: SIZES.buttonRadius,
    backgroundColor: COLORS.backgroundGrey,
    alignSelf: 'center',
    alignItems: 'center',
    justifyContent: 'center',
    ...SHADOW
  },
  updateText: {
    fontFamily: 'MontserratBold',
    color: COLORS.white,
    fontSize: SIZES.buttonText * 0.9
  },
  accordionsBox: {
    width: '100%',
    alignSelf: 'center'
  },
  sectionContainer: {
    width: '100%',
    justifyContent: 'center',
    marginVertical: 5,
    backgroundColor: COLORS.white
  },
  section: {
    // flexGrow: 1
  },
  sectionName: {
    fontFamily: 'MontserratBold',
    fontSize: SIZES.p18,
    lineHeight: 40,
    paddingLeft: 10,
    backgroundColor: COLORS.cream,
    marginBottom: 3
  },

  sectionContentBox: {
    width: '100%',
    backgroundColor: COLORS.white,
    borderRadius: 10
  },
  skinTypesButton: {
    ...BORDER,
    height: 40,
    alignItems: 'center',
    justifyContent: 'center',
    width: '90%',
    alignSelf: 'center'
  },
  skinType: {
    fontSize: SIZES.p18,
    fontFamily: 'Montserrat'
  },
  settingInputs: {
    justifyContent: 'center',
    alignSelf: 'center',
    width: '90%',
    paddingLeft: 10,
    height: 40,
    marginBottom: 3,
    borderRadius: SIZES.buttonRadius,
    ...BORDER,
    fontFamily: 'MontserratBold'
  }
})

export default styles
