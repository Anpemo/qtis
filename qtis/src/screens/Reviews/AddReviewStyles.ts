
import { COLORS, SIZES } from '../../../constants'
import { StyleSheet } from 'react-native'
const AddReviewStyles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.white
  },
  backgroundImage: {
    flex: 1,
    resizeMode: 'cover',
    width: '100%'
  },
  secondContainer: {
    flex: 1,
    alignItems: 'center',
    width: SIZES.width,
    justifyContent: 'center'
  },
  backIcon: {
    color: COLORS.black,
    marginBottom: 5,
    marginLeft: 10,
    position: 'absolute',
    top: 40,
    left: 40
  },
  inputsContainer: {
    width: SIZES.width * 0.8,
    height: '60%',
    alignSelf: 'center',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: SIZES.squareRadius,
    backgroundColor: COLORS.categoryTextBackground
  },
  rateInput: {
    width: '90%',
    height: '20%',
    margin: 5,
    justifyContent: 'center',
    backgroundColor: COLORS.cream,
    fontSize: SIZES.p20,
    color: COLORS.lettersGrey,
    fontFamily: 'Montserrat',
    padding: 10
  },
  textInput: {
    width: '90%',
    height: '70%',
    margin: 5,
    justifyContent: 'center',
    backgroundColor: COLORS.cream,
    fontSize: SIZES.p20,
    color: COLORS.lettersGrey,
    fontFamily: 'Montserrat',
    padding: 10
  },
  button: {
    alignSelf: 'center',
    width: SIZES.width * 0.8,
    justifyContent: 'center',
    backgroundColor: COLORS.black,
    marginTop: 15,
    height: SIZES.buttonheight,
    borderRadius: SIZES.buttonRadius
  },
  buttonText: {
    color: COLORS.white,
    textAlign: 'center',
    fontSize: SIZES.buttonText * 0.8,
    fontFamily: 'Montserrat'
  }

})

export default AddReviewStyles
