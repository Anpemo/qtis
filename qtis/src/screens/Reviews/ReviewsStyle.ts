
import { COLORS, SIZES, REVIEWBORDER } from '../../../constants'
import { StyleSheet } from 'react-native'
const ReviewsStyle = StyleSheet.create({
  container: {
    flex: 1,
    width: '100%'
  },
  filterContainer: {
    width: '100%',
    alignItems: 'flex-start',
    justifyContent: 'center',
    marginVertical: 10,
    paddingLeft: 10
  },
  filterTitle: {
    fontSize: SIZES.p18,
    color: COLORS.black,
    fontFamily: 'Montserrat'
  },
  filtersBox: {
    flexDirection: 'row',
    width: '100%'
  },
  filterButton: {
    width: 130,
    height: 30,
    backgroundColor: COLORS.cream,
    borderRadius: SIZES.buttonRadius,
    alignItems: 'center',
    justifyContent: 'center',
    margin: 3
  },
  filterText: {
    fontFamily: 'Montserrat',
    fontSize: SIZES.p16
  },
  reviewsContainer: {
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center'
  },
  valorationContainer: {
    flexDirection: 'row',
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    marginBottom: 3
  },
  punctuation: {
    fontSize: SIZES.p20,
    fontFamily: 'Montserrat',
    color: COLORS.black,
    marginRight: 3,
    lineHeight: 20
  },
  reviewTitleContainer: {
    width: '95%',
    paddingTop: 15,
    marginVertical: 10,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  reviewsTitle: {
    fontSize: SIZES.p20,
    fontFamily: 'MontserratBold'
  },
  reviewsFlatList: {
    width: '100%',
    marginBottom: 50,
    marginTop: 10,
    alignItems: 'center'
  },
  reviewBox: {
    width: '95%',
    flexDirection: 'row',
    marginBottom: 10
  },
  userPictureBox: {
    alignSelf: 'flex-start',
    justifyContent: 'center',
    alignContent: 'center',
    marginRight: 10
  },
  userPicture: {
    height: 70,
    width: 70,
    resizeMode: 'cover',
    borderRadius: 100,
    backgroundColor: COLORS.cream
  },
  reviewContainer: {
    flex: 1,
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
    padding: 10,
    borderRadius: SIZES.squareRadius,
    ...REVIEWBORDER
  },
  userName: {
    fontSize: SIZES.p18,
    color: COLORS.brown,
    fontFamily: 'MontserratBold'
  },
  reviewText: {
    fontSize: SIZES.p18,
    color: COLORS.brown,
    fontFamily: 'Montserrat',
    backgroundColor: COLORS.white
  }
})

export default ReviewsStyle
