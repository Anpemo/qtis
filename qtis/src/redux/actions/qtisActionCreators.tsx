import qtisActionTypes from './qtisActionTypes'
import axios from 'axios'
import serverUrls from '../../../constants/serverUrls'
import UserInterface from '../../Interfaces/UserInterface'
import ProductInterface from '../../Interfaces/ProductInterface'
import ReviewInterface from '../../Interfaces/ReviewInterface'
import { Dispatch } from 'react'
import { userRegisterAction, updateUserAction, createReviewAction, userLoginAction, fetchProductsAction, fetchProductAction, fetchReviewsAction, createProductAction } from '../../models'

export function userRegister (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<userRegisterAction>) {
    try {
      const { data } = await axios.post(serverUrls.BACKEND_REGISTER, userData)

      dispatch({
        type: qtisActionTypes.USER_REGISTER,
        data
      })
    } catch {
      dispatch({
        type: qtisActionTypes.USER_REGISTER,
        data: 401
      })
    }
  }
}

export function userLogin (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<userLoginAction>) {
    try {
      const { data } = await axios.post(serverUrls.BACKEND_LOGIN, userData)
      dispatch({
        type: qtisActionTypes.USER_LOGIN,
        data
      })
    } catch {
      dispatch({
        type: qtisActionTypes.USER_LOGIN,
        data: 401
      })
    }
  }
}

export function fetchProducts (category: Object) {
  return async function fetchInfo (dispatch: Dispatch<fetchProductsAction>) {
    const { data } = await axios.get(`${serverUrls.PRODUCTS}${category}`)
    dispatch({
      type: qtisActionTypes.PRODUCTS_LIST,
      data
    })
  }
}

export function fetchProduct (barCodeData: String) {
  return async function fetchInfo (dispatch: Dispatch<fetchProductAction>) {
    let date
    try {
      const { data, headers } = await axios.get(`${serverUrls.PRODUCTS}${barCodeData}`)
      date = headers.date
      dispatch({
        type: qtisActionTypes.SINGLE_PRODUCT,
        data: data[0]
      })
    } catch {
      dispatch({
        type: qtisActionTypes.SINGLE_PRODUCT,
        data: { data: 'Product does not exist', date }
      })
    }
  }
}

export function createProduct (productData: ProductInterface) {
  return async function fetchInfo (dispatch: Dispatch<createProductAction>) {
    const { data } = await axios.post(serverUrls.PRODUCTS, productData)
    dispatch({
      type: qtisActionTypes.CREATE_PRODUCT,
      data
    })
  }
}

export function cleanProduct () {
  return ({
    type: qtisActionTypes.CLEAN_PRODUCT,
    data: null
  })
}

export function fetchReviews (parameter: Object) {
  return async function fetchInfo (dispatch: Dispatch<fetchReviewsAction>) {
    const { data } = await axios.get(`${serverUrls.REVIEWS}${parameter}`)

    dispatch({
      type: qtisActionTypes.REVIEWS_LIST,
      data
    })
  }
}

export function createReview (reviewData: ReviewInterface) {
  return async function fetchInfo (dispatch: Dispatch<createReviewAction>) {
    const { data } = await axios.post(serverUrls.REVIEWS, reviewData)
    dispatch({
      type: qtisActionTypes.CREATE_REVIEW,
      data
    })
  }
}

export function updateUser (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<updateUserAction>) {
    const { data } = await axios.put(serverUrls.UPDATE_USER, userData)
    dispatch({
      type: qtisActionTypes.UPDATE_USER,
      data
    })
  }
}
