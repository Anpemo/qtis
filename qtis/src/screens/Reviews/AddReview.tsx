import React, { useState } from 'react'
import {
  TouchableOpacity, View, Text, ImageBackground, KeyboardAvoidingView
} from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { images } from '../../../constants'
import styles from './AddReviewStyles'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { createReview } from '../../redux/actions/qtisActionCreators'
import { AntDesign } from '@expo/vector-icons'
import { TextInput } from 'react-native-gesture-handler'

function AddReview ({ navigation, route, actions, user }: any) {
  const { productBarCode } = route.params
  const { userName, userPicture, _id } = user
  const { skinType } = user
  const [rating, setRating] = useState('')
  const [reviewText, setReviewText] = useState('')
  function shareReview () {
    actions.createReview({ rating: rating.replace(/,/g, '.'), reviewText, productBarCode, userId: _id, userName, userPicture, skinType })
    navigation.navigate('ProductDetail', productBarCode)
  }
  return (
    <SafeAreaView style={styles.container}>
      <ImageBackground source={images.backgroundReview} style={styles.backgroundImage}>
      <View style={styles.secondContainer}>
      <TouchableOpacity
      onPress={() => { navigation.goBack() }}
      style={styles.backIcon} >
        <AntDesign name="doubleleft" size={22}/>
      </TouchableOpacity>
        <KeyboardAvoidingView behavior={'padding'}>
      <View style={styles.inputsContainer}>
      <TextInput
          onChangeText={(number) => setRating(number.toString())}
          placeholder={'Rate from 1 to 5'}
          style={styles.rateInput}
          value={rating}
          keyboardType={'decimal-pad'}

        />
        <TextInput
          onChangeText={(text) => setReviewText(text)}
          placeholder={'Write your review'}
          style={styles.textInput}
          value={reviewText}
          multiline={true}
          numberOfLines={20}
        />
      </View>
        <TouchableOpacity
        onPress={() => shareReview()}
        style={styles.button}
        >
            <Text style={styles.buttonText}>SHARE</Text>
        </TouchableOpacity>
          </KeyboardAvoidingView>
      </View>
      </ImageBackground>
    </SafeAreaView>
  )
}
function mapStateToProps ({ productsReducer, userReducer }: any) {
  return {
    product: productsReducer.product,
    user: userReducer.user
  }
}

function mapDispatchToProps (dispatch: any) {
  return {
    actions: bindActionCreators({
      createReview
    }, dispatch)
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(AddReview)
