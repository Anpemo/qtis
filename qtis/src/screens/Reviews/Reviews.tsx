import React, { useEffect } from 'react'
import {
  View, Text, FlatList, Image
} from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { Rating } from 'react-native-ratings'
import { connect } from 'react-redux'
import { fetchReviews } from '../../redux/actions/qtisActionCreators'
import { bindActionCreators } from 'redux'
import styles from './ReviewsStyle'

function Reviews ({ reviews, actions, route, parameter }: any) {
  console.log('ruta', parameter)

  useEffect(() => {
    actions.fetchReviews(parameter)
  }, [])

  function renderReviews ({ item }: any) {
    (
        <View style={styles.reviewBox}>
          <View style={styles.userPictureBox}>
            <Image
            source={reviews.userPicture}
            style={styles.userPicture}
            key={1}
            />
          </View>
          <View style={styles.reviewContainer}>
            <Text style={styles.userName}>Angela Pedrero
            </Text>
            <Text style={styles.reviewText}>{item.reviewText}
            </Text>
          </View>
        </View>
    )
  }
  return (
        <SafeAreaView>
            <View style={styles.reviewsContainer}>
              <View style={styles.valorationContainer}>
                <Text style={styles.punctuation}>4,5</Text>
                <Rating
                type='star'
                ratingCount={5}
                readonly={true}
                fractions={4}
                imageSize={30}
                startingValue={4.5}
                ratingBackgroundColor={'black'}
                />
              </View>
              <FlatList
                  data = {reviews}
                  renderItem={renderReviews}
                  keyExtractor = {(item: any) => item.id}
                  horizontal={false}
                  style={styles.reviewsFlatList}

                  />
              </View>
        </SafeAreaView>
  )
}
function mapStateToProps ({ reviewsReducer }: any) {
  return {
    product: reviewsReducer.product
  }
}

function mapDispatchToProps (dispatch: any) {
  return {
    actions: bindActionCreators({
      fetchReviews
    }, dispatch)
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(Reviews)
