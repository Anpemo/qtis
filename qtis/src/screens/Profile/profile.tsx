import React, { useEffect, useState } from 'react'
import {
  Image, View, Text, ImageBackground, KeyboardAvoidingView, TouchableOpacity, ScrollView, Platform, TextInput
} from 'react-native'
import { images } from '../../../constants'
import { SafeAreaView } from 'react-native-safe-area-context'
import skinTypes from '../../../constants/skinTypes'
import styles from './ProfileStyles'
import * as ImagePicker from 'expo-image-picker'
import { Ionicons } from '@expo/vector-icons'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import Reviews from '../../../src/screens/Reviews/Reviews'
import { updateUser } from '../../redux/actions/qtisActionCreators'

function Profile ({ user, actions }: any) {
  const [openSkinType, setOpenSkinType] = useState(false)
  const [openReviews, setOpenReviews] = useState(false)
  const [openSettings, setOpenSettings] = useState(false)
  const [userPicture, setUserPicture] = useState(user.userPicture)
  const [skinType, setSkinType] = useState(user.skinType)
  const [userName, setUserName] = useState(user.userName)
  const [age, setAge] = useState(user.age)
  const [city, setCity] = useState(user.city)

  useEffect(() => {
    (async () => {
      if (Platform.OS !== 'web') {
        const { status } = await ImagePicker.requestMediaLibraryPermissionsAsync()
        if (status !== 'granted') {
          alert('Sorry, we need camera roll permissions to make this work!')
        }
      }
    })()
  }, [])

  const pickImage = async () => {
    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.All,
      allowsEditing: true,
      aspect: [4, 3],
      quality: 1
    })

    if (!result.cancelled) {
      setUserPicture(result.uri)
    }
  }
  function selectSkinType (category : any) {
    setSkinType(category)
    setOpenSkinType(false)
  }
  return (
    <KeyboardAvoidingView behavior='padding' style={{ flex: 1, paddingBottom: 50 }}>
    <SafeAreaView style={styles.container}>
      <ScrollView style={styles.scroll}>
      <ImageBackground source={images.profileHeader} style={styles.header} >
          <View style={styles.pictureBox}>
            { userPicture
              ? <Image source={{ uri: userPicture }} style={styles.userPicture} />
              : <Text style={styles.userPictureText}>ADD A PICTURE</Text>}
            <View style={styles.cameraButtonContainer} >
              <Ionicons name="camera-outline" size={35} color="grey" onPress={pickImage}/>
            </View>
          </View>
      </ImageBackground>
      <View style={styles.body}>
        <View style={styles.userInformation}>
          <Text style={styles.userName}>{user.userName}
          </Text>
          {user.age &&
            <Text style={styles.userData}>{`${user.age} years old`}
            </Text>
          }
          <Text style={styles.userData}>{user.city}
          </Text>
        </View>
      <View style={styles.accordionsBox}>
        <TouchableOpacity
        onPress={() => {
          setOpenSkinType(!openSkinType)
        }}
        style={styles.sectionContainer}
        activeOpacity={0.5}
        >
            <View style={styles.section}>
              <Text style={styles.sectionName}>SKIN TYPE: {skinType.toUpperCase()}</Text>
              <View style={styles.sectionContentBox}>
                {openSkinType && skinTypes.map((skinType) => (
                  <TouchableOpacity
                  key={skinType}
                  activeOpacity={0.5}
                  style={styles.skinTypesButton}
                  onPress={() => selectSkinType(skinType)}
                  >
                    <Text style={styles.skinType}>{skinType.toUpperCase()}</Text>
                  </TouchableOpacity>
                ))}
              </View>
            </View>

        </TouchableOpacity>

        <TouchableOpacity
        onPress={() => {
          setOpenReviews(!openReviews)
        }}
        style={styles.sectionContainer}
        activeOpacity={0.5}
        >
            <View style={styles.section}>
              <Text style={styles.sectionName}>YOUR REVIEWS</Text>
              <View style={styles.sectionContentBox}>
                {openReviews &&
                  <Reviews parameter={user._id}/> }
              </View>
            </View>

        </TouchableOpacity>

        <TouchableOpacity
        onPress={() => {
          setOpenSettings(!openSettings)
        }}
        style={styles.sectionContainer}
        activeOpacity={0.5}
        >
            <View style={styles.section}>
              <Text style={styles.sectionName}>ACCOUNT SETTINGS</Text>
              <View >
                {openSettings &&
                  <View>

                    <TextInput
                      onChangeText={(event) => setUserName(event)}
                      placeholder={'What\'s your name?'}
                      style={styles.settingInputs}
                      />
                    <TextInput
                    placeholder={'How old are you?'}
                    onChangeText={(event) => setAge(event.toString())}
                    style={styles.settingInputs}
                    keyboardType={'decimal-pad'}
                    />

                      <TextInput
                      placeholder={'Where are you from?'}
                      onChangeText={(event) => setCity(event)}
                      style={styles.settingInputs}
                      />
                    <TouchableOpacity style={styles.updateButton}>
                      <Text style={styles.updateText} onPress={() => actions.updateUser({ city, age, userName, _id: user._id, userPicture, skinType })}>UPDATE YOUR DATA</Text>
                    </TouchableOpacity>
                  </View>
                }
              </View>
            </View>

        </TouchableOpacity>
      </View>
      </View>
    </ScrollView>
    </SafeAreaView>
  </KeyboardAvoidingView>
  )
}
function mapStateToProps ({ userReducer }: any) {
  return {
    user: userReducer.user
  }
}

function mapDispatchToProps (dispatch: any) {
  return {
    actions: bindActionCreators({
      updateUser
    }, dispatch)
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(Profile)
