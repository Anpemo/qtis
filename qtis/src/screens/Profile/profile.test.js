import { cleanup, fireEvent, render } from '@testing-library/react-native'
import Profile from './profile'
import renderer from 'react-test-renderer'
import React, { useState } from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import * as actions from '../../redux/actions/qtisActionCreators'
import Reviews from '../Reviews/Reviews'
// import * as ImagePicker from '../ImagePicker'
jest.mock('../../redux/actions/qtisActionCreators')
const mockedGoBack = jest.fn()
const mockedNavigate = jest.fn()

jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => ({
      goBack: mockedGoBack,
      navigate: mockedNavigate
    })
  }
})
describe('Given a Profile component', () => {
  let component
  const navigate = jest.fn()
  const goBack = jest.fn()
  const params = 123123123

  beforeEach(() => {
    const mockStore = configureStore()

    component = (
          <Provider store={mockStore({
            userReducer: { user: { age: 14 }, userPicture: 'asd' },
            reviewsReducer: {
              reviews: []
            }
          }
          )}><Profile navigation={{ navigate, goBack }} route={{ params }} /></Provider>
    )
  })

  afterEach(() => {
    cleanup()
  })

  describe('When opening reviews', () => {
    test('Then it will match the snapshot', () => {
      jest.spyOn(actions, 'fetchReviews').mockReturnValue({ type: '' })

      const { getByTestId } = render(component)
      const button = getByTestId('openReviews')
      fireEvent.press(button)

      const mockReviewsStore = configureStore()
      const reviewComponent = (<Provider store= { mockReviewsStore({
        reviewsReducer: {
          reviews: []
        }
      })}>
          <Reviews route={{ params }} />
        </Provider>
      )

      const componentRendered = render(reviewComponent)
      expect(componentRendered).toMatchSnapshot()
    })
  })

  describe('When oppening account settings, writing on inputs and pressing action.updateUser', () => {
    test('Then action.updateUser will be called with { city, age, userName, _id, userPicture, skinType } ', () => {
      const city = 'Barcelona'
      const age = 26
      const userName = 'Angela'
      const _id = 'asd423af'
      const userPicture = 'asdasdas'
      const skinType = 'combination'
      jest.spyOn(actions, 'updateUser').mockReturnValue({ type: '' })
      const { getByPlaceholderText, getByTestId } = render(component)

      const openSettingsButton = getByTestId('openSettings')
      fireEvent.press(openSettingsButton)
      // write on userName input:
      const userNameInput = getByPlaceholderText('What\'s your name?')
      const newUserName = 'newValue'
      fireEvent.changeText(userNameInput, newUserName)
      // write on age input:
      const ageInput = getByPlaceholderText('How old are you?')
      const newAge = 'newAge'
      fireEvent.changeText(ageInput, newAge)
      // write on city input:
      const cityInput = getByPlaceholderText('Where are you from?')
      const newCity = 'newCity'
      fireEvent.changeText(cityInput, newCity)
      // find function and press it
      fireEvent.press(getByTestId('updateUserButton'))
      expect(actions.updateUser).toHaveBeenCalled()
    })
  })
  describe('When oppening skin type and pressing on a skintype', () => {
    test('Then setOpenSkinType will be called with false', () => {
      const { getByTestId, getAllByTestId } = render(component)
      const button = getByTestId('openSkinType')
      fireEvent.press(button)
      const skinType = getAllByTestId('skinTypesMap')[0]
      fireEvent.press(skinType)
      expect(skinType).toHaveBeenCalled()
    })
  })
})
