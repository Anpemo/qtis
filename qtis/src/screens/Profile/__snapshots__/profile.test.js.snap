// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Profile component When opening reviews Then it will match the snapshot 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginVertical": 10,
          "paddingTop": 15,
          "width": "95%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontFamily": "MontserratBold",
            "fontSize": 20,
          }
        }
      >
        REVIEWS
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "marginBottom": 50,
          "marginTop": 10,
          "width": "100%",
        }
      }
    />
  </View>
</RNCSafeAreaView>
`;
